### PR TITLE
Feature: new firmware.py flasher

### PR DESCRIPTION
**Pull request by krautech**
## Checklist

The following relevant macros have been tested:

- [ ] `CARTOGRAPHER_CALIBRATE`
- [ ] `PROBE`
- [ ] `PROBE_ACCURACY`
- [ ] `CARTOGRAPHER_THRESHOLD_SCAN`
- [ ] `CARTOGRAPHER_TOUCH CALIBRATE=1`
- [ ] `CARTOGRAPHER_TOUCH`
- [ ] `CARTOGRAPHER_TOUCH METHOD=manual`
- [ ] `BED_MESH_CALIBRATE`

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [tool.pyright]
 pythonVersion = "3.8"
-include = ["scanner.py"]
+include = ["scanner.py","scripts/**.py"]
```
Comment (Jomik) [inline pyproject.toml:3]:
```suggestion
include = ["scanner.py", "scripts/**.py"]
```

File: pyproject.toml [hunk 2]
```diff
@@ -10,4 +10,4 @@ failOnWarnings = false
 
 [tool.ruff]
 target-version = "py38"
-include = ["scanner.py"]
+include = ["scanner.py","scripts/**.py"]
```
Comment (Jomik) [inline pyproject.toml:13]:
```suggestion
include = ["scanner.py", "scripts/**.py"]
```

File: scripts/firmware.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    YELLOW = "\033[93m"
+    MAGENTA = "\033[95m"
+    CYAN = "\033[96m"
+
+    def colored_text(self, text: str, color: str) -> str:
+        return f"{color}{text}{self.RESET}"
+
+
+class Menu:
+    def __init__(self, title: str, menu_items: dict):
```
Comment (Jomik) [inline scripts/firmware.py:24]:
```suggestion
    def __init__(self, title: str, menu_items: dict: [str, MenuItem]):
```
Make a MenuItem class? 😄

File: scripts/firmware.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    def display(self):
+        while True:
+            # Print menu header
+            print("=" * 40)
+            print(self.color.colored_text(self.title, self.color.RED))
+            print("=" * 40)
+
+            # Print menu items
+            for key, (description, _) in self.menu_items.items():
+                print(f"{key}. {description}")
```
Comment (Jomik) [inline scripts/firmware.py:38]: What happens if `key` is "0" ?

File: scripts/firmware.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+                    f"{Color.GREEN}You selected: {self.menu_items[choice][0]}{Color.RESET}\n"
+                )
+                function()  # Call the associated function
+            else:
+                print(f"{Color.RED}Invalid choice. Please try again.{Color.RESET}")
+
+
+class Firmware:
+    def __init__(self) -> None:
+        self.color = Color()
```
Comment (Jomik) [inline scripts/firmware.py:67]: This should be redundant.
```suggestion

```

File: scripts/firmware.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
```
Comment (Jomik) [inline scripts/firmware.py:68]:
```suggestion
```

File: scripts/firmware.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    def error_msg(self, message: str) -> None:
+        if message is not None and DEBUG:
+            print(self.color.colored_text("Error:", self.color.RED), message)
```
Comment (Jomik) [inline scripts/firmware.py:71]:
```suggestion
            print(self.color.colored_text("Error:", Color.RED), message)

```

File: scripts/firmware.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+    def find_can_uuid(self) -> None:
+        self.step_title("Finding CAN Device UUID")
+
+    # Check status of UUID
+    def get_can_status(self, uuid: str) -> None:
+        self.step_title("Checking CAN Device Status")
+        if not os.path.exists(KLIPPER_DIR):
+            return self.error_msg(f"Directory '{KLIPPER_DIR}' does not exist")
+        if uuid is None:
+            return
```
Comment (Jomik) [inline scripts/firmware.py:108]: This doesn't make sense, but guess it's a work in progress?
Can uuid be `None`? Not according to the signature 😄

File: scripts/firmware.py [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+    def header(self):
+        print("=" * 100)
+        print(
+            " " * 35
+            + self.color.colored_text("CARTOGRAPHER FIRMWARE FLASHER", self.color.CYAN)
+        )
+        print("=" * 100)
+
+    # find can uuid from klippy.log
+    def find_can_uuid(self) -> None:
```
Comment (Jomik) [inline scripts/firmware.py:99]: Does this fit on the `Firmware` class?

**Review by Jomik — CHANGES_REQUESTED** [review]

File: scripts/firmware.py (hunk 8 above)
Comment (krautech) [reply scripts/firmware.py:108]: haha def work in progress :D

File: scripts/firmware.py (hunk 9 above)
Comment (krautech) [reply scripts/firmware.py:99]: Put it in its own? but what to name it?

File: scripts/firmware.py (hunk 3 above)
Comment (krautech) [reply scripts/firmware.py]: Hows it look now?

File: scripts/firmware.py (hunk 4 above)
Comment (krautech) [reply scripts/firmware.py]: Fixed i think?

**Comment by krautech** [comment]
@Jomik feel free to comment, happy to make changes etc. Its now 7am so i probably messed up somewhere but it flashes OMG so well and fast. I'll do USB and DFU pending your thoughts.

File: scripts/firmware.py [hunk 10]
```diff
@@ -0,0 +1,1086 @@
...
+        "--flash",
+        help="Specify the flashing mode (CAN, USB or DFU)",
+        required=False,
+        default=None,  # Default value if -f is not provided
+        choices=["CAN", "USB", "DFU"],  # Restrict to these options
+        type=lambda s: s.upper(),  # Convert input to uppercase
+    )
+
+    # Parse the arguments
+    args = parser.parse_args()
```
[the remark this answers is not in the document]
Comment (krautech) [reply scripts/firmware.py:1060]: I'll have a play with this. It wasn't so bad before I added a ton of args 😂

File: scripts/firmware.py [hunk 11]
```diff
@@ -0,0 +1,1086 @@
+import os
+import re
+import subprocess
+import argparse
+import shutil
+import tempfile
+import fnmatch
+
+from packaging.version import Version
+from enum import StrEnum  # type: ignore
```
[the remark this answers is not in the document]
Comment (krautech) [reply scripts/firmware.py:10]: Well that's annoying.. I'll work on it. Sigh

File: scripts/firmware.py [hunk 12]
```diff
@@ -0,0 +1,1086 @@
+import os
+import re
+import subprocess
+import argparse
+import shutil
+import tempfile
+import fnmatch
+
+from packaging.version import Version
```
[the remark this answers is not in the document]
Comment (krautech) [reply scripts/firmware.py:9]: I don't know if it'll be as possible. Need something to sort through versioning on subdirectories of the firmware folders hmm

File: scripts/firmware.py [hunk 13]
```diff
@@ -0,0 +1,1086 @@
...
+            )  # Relative path of the subdirectory
+
+            if high_temp:
+                # If high_temp is True, process only "HT" directories
+                if "HT" not in subdirectory:
+                    continue
+            else:
+                # If high_temp is False, skip "HT" directories
+                if "HT" in subdirectory:
+                    continue
```
[the remark this answers is not in the document]
Comment (krautech) [reply scripts/firmware.py:710]: For some reason when I tried this, it wouldn't play nice. It was also like 9am with no sleep, so.

**Comment by krautech** [comment]
I'll try split the functions up more, especially within `Firmware` where I think most of the clutter is.

File: scripts/firmware.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (krautech) [reply scripts/firmware.py:10]: Should now be resolved

File: scripts/firmware.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (krautech) [reply scripts/firmware.py:9]: Should now be resolved

File: scripts/firmware.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (krautech) [reply scripts/firmware.py:1083]: Should be done :)

File: scripts/firmware.py [hunk 14]
```diff
@@ -0,0 +1,1086 @@
...
+                    command, text=True, capture_output=True, check=True
+                )
+
+                # Output the results
+                self.flash_success(result.stdout.strip())
+            except subprocess.CalledProcessError as e:
+                # Handle errors during the command execution
+                self.flash_fail(f"Error flashing firmware: {e}")
+                if e.stderr:
+                    print(e.stderr.strip())
```
[the remark this answers is not in the document]
Comment (krautech) [reply scripts/firmware.py:983]: done

**Pull request closed without merging**